tools/reloc_preq.c

```diff
@@ -0,0 +1,60 @@
+#include <stddef.h>
+#include <stdio.h>
+#include <string.h>
+
+#include "spec.h"
+#include "util.h"
+
+void print_usage(char* prog_name) {
+    printf("USAGE: %s SPEC OVERLAY_SEGMENT_NAME\n"
+           "Search the preprocessed spec for an overlay segment name, \n"
+           "e.g. \"ovl_En_Firefly\", and return a list of the files it includes\n",
+           prog_name);
+}
+
+int main(int argc, char** argv) {
+    char* spec_path;
+    char* overlay_name;
+    char* spec;
+    size_t size;
+    Segment* segments = NULL;
+    int segments_count = 0;
+    Segment found_segment = { 0 };
+
+    if (argc != 3) {
+        print_usage(argv[0]);
+        return 1;
+    }
+
+    spec_path = argv[1];
+    overlay_name = argv[2];
+
+    // printf("spec path: %s\n", spec_path);
+    // printf("overlay name: %s\n", overlay_name);
+
+    spec = util_read_whole_file(spec_path, &size);
+    parse_rom_spec(spec, &segments, &segments_count);
+
+    {
+        int i;
+        for (i = 0; i < segments_count; i++) {
+            if (strcmp(segments[i].name, overlay_name) == 0) {
+                found_segment = segments[i];
+                break;
+            }
+        }
+        if (i == segments_count) {
+            fprintf(stderr, "No segment \"%s\" found.\n", overlay_name);
+            return 1;
+        }
+    }
+    {
```

Comment:
1. I usually do this to get the best of both worlds: it's like having a transparent inline function without having to make all the arguments and so on. It keeps the code nicely separated without having to restructure it a lot. It's also a bit presumptuous to assume that anyone round here knows modern C, when most of us were brought up on C89 lol.
2. Will change.
3. Oh, that needs proper documentation, then... I'm pretty sure I just looked at the header and the other uses.
4. I suppose this is worth it, since it has pointers to strings in it, so the spec can't be freed sooner if it isn't a pointer anyway.

Reply:
(1) fwiw tools already rely on not-C89, it's mostly C89-like out of habit I guess but there's at least this
https://github.com/zeldaret/oot/blob/cd1d08d34f32ec97396786e83783fc37b7a68858/tools/spec.c#L126

(4) `Segment#includes` is a pointer so you would need to deepcopy then. It sounds less bug-friendly to just use a pointer and free the whole thing at the end
Also I just noticed reading spec.c, the strings aren't copied but only pointers to inside the spec text returned by `util_read_whole_file`, so freeing the segments should be conceptually done before freeing the spec